### PR TITLE
test: add missing analyzer unit tests and datafinder date validation

### LINKED ISSUE #3782: Resolve inline TODOs: missing analyzer tests and datafinder date validation
Hi team 👋
While exploring the codebase, I noticed a few places where maintainers had left 
`TODO` comments requesting test coverage and input validation that were never 
followed up on. I'd like to address all of them in a single focused PR.
### What's currently missing
| File | Issue |
|------|-------|
| `timesketch/lib/analyzers/ntfs_timestomp_test.py` | Test body is `self.assertEqual(True, True)` — no real assertions |
| `timesketch/lib/analyzers/domain_test.py` | Only 22 lines, tests class instantiation only |
| `timesketch/lib/analyzers/feature_extraction_test.py` | Has `# TODO: Add tests for the feature extraction` with nothing below it |
| `timesketch/lib/datafinder.py` | Two instances of `# TODO: Implement a check if this is a valid ISO formatted date` |
### Proposed fix
- Replace the NTFS timestomp stub with real behavioral tests covering all branches 
  of `is_suspicious()`.
- Expand the domain analyzer test file with MockDataStore-based tests (empty stream, 
  URL extraction, rare domains, CDN detection).
- Add unit tests for `FeatureExtractionSketchPlugin` — properties, empty plugin name, 
  and unregistered plugin error handling.
- Implement ISO 8601 validation in `set_start_date()` / `set_end_date()` using 
  `datetime.fromisoformat()`. Validation is non-breaking: invalid dates still get 
  accepted with a logged warning to preserve backward compatibility.

Happy to discuss any of this before I open the PR. Thanks!

### PR DESCRIPTION
**Pull request by SNO7E-G**
Hi Team 👋
This PR cleans up four `TODO` comments that have been sitting in the codebase for 
a while — all of them asking for test coverage or input validation that was never 
implemented. Nothing drastic, just filling in the gaps.
Closes #3782
---
### What's been bothering me (and hopefully you too)
The NTFS timestomp test file had this as its only test:
    self.assertEqual(True, True)
The domain test file was 22 lines and only checked that the class could be 
instantiated. The feature extraction test file had a `TODO` comment and nothing 
else. And `DataFinder` was silently accepting any string as a date without 
ever checking if it was actually ISO 8601.
### What this PR does
**`ntfs_timestomp_test.py`** — 5 new behavioral tests for `is_suspicious()`:
- No `std_info_event` → returns False
- Empty `file_names` → returns False  
- Timestamps within threshold → not flagged, no attributes set
- Timestamps above threshold → flagged, `time_delta` / `time_deltas` set correctly
- Mixed file names (one within threshold) → detection correctly suppressed
**`domain_test.py`** — 4 new MockDataStore-backed behavioral tests:
- Empty event stream → graceful early exit with `"No domains to analyze."`
- Events with `url` field (no `domain`) → correct domain extraction
- Rare domain threshold → domain counted in result summary
- Known CDN domain (`.cloudfront.net`) → reported in CDN network count
**`feature_extraction_test.py`** — 3 new tests for `FeatureExtractionSketchPlugin`:
- Getters and setters work for `plugin_name`, `feature_name`, `feature_config`
- Empty `plugin_name` → returns `"Feature extraction plugin name is empty"` cleanly
- Unregistered plugin name → gracefully returns wrapped `ValueError` message
**`datafinder.py`** — ISO 8601 date validation:
- Added `_is_valid_iso_date()` helper using `datetime.fromisoformat()`
- `set_start_date()` and `set_end_date()` now log a warning if the input isn't valid
- Non-breaking: the value is still set regardless, so no existing callers are affected
- Removes both `TODO` comments
---
### Checks
- [x] All tests pass
- [x] Unit tests added
- [x] e2e tests added — *not applicable; changes are pure unit-level with no new endpoints or UI behavior*
- [ ] Documentation updated — *not applicable; no public API or user-facing behavior changed*
**Closing issues**
Closes #3782

### CODE REVIEW

**Comment by SNO7E-G** [comment]
@jaegeral
Done! I've fixed all 3 CI failures and pushed the corrected commit

File: timesketch/lib/analyzers/ntfs_timestomp_test.py [hunk 1]
```diff
@@ -108,9 +108,85 @@ def test_is_suspicious(self):
             self.assertEqual(std_diffs, tc.expected_si_diffs)
             self.assertEqual(fn_diffs, tc.expected_fn_diffs)
 
-    # Mock the OpenSearch datastore.
     @mock.patch("timesketch.lib.analyzers.interface.OpenSearchDataStore", MockDataStore)
-    def test_analyzer(self):
-        """Test analyzer."""
-        # TODO: Write actual tests here.
```
Comment (jkppr) [inline timesketch/lib/analyzers/ntfs_timestomp_test.py:115]: The original TODO was sitting under a placeholder `test_analyzer(self)`. The goal of that TODO was to test the analyzer's main `run()` method (which queries the datastore, groups events, and creates the Timesketch view), rather than just the `is_suspicious()` helper.

Currently, the `run()` method is completely untested. Would you be open to helping us write an integration test for it? It would look something like this:

* Populating the `MockDataStore` with a couple of mock events (one with `attribute_type: 16` and one with `48`).
* Running `analyzer.run()`.
* Verifying that the events are updated/committed and the `NtfsTimestomp` view is added.

File: timesketch/lib/analyzers/ntfs_timestomp_test.py [hunk 2]
```diff
@@ -108,9 +108,85 @@ def test_is_suspicious(self):
...
+            file_reference=1,
+            timestamp_desc="Content Modification Time",
+            std_info_event=MockEvent(),
+            std_info_timestamp=0,
+            file_names=[],
+        )
+        self.assertFalse(analyzer.is_suspicious(file_info))
+
+    @mock.patch("timesketch.lib.analyzers.interface.OpenSearchDataStore", MockDataStore)
+    def test_is_suspicious_within_threshold(self):
```
Comment (jkppr) [inline timesketch/lib/analyzers/ntfs_timestomp_test.py:138]: Isn't this already covered by the existing `test_is_suspicious` test? (Case 1, 4, and 5 - The existing tests already thoroughly prove that any `analyzer.threshold` (which includes `analyzer.threshold - 1`) is correctly ignored.)

File: timesketch/lib/analyzers/ntfs_timestomp_test.py [hunk 3]
```diff
@@ -108,9 +108,85 @@ def test_is_suspicious(self):
...
+            std_info_event=MockEvent(),
+            std_info_timestamp=0,
+            file_names=[(fn_event, analyzer.threshold - 1)],
+        )
+        self.assertFalse(analyzer.is_suspicious(file_info))
+        # The time_delta attribute must NOT have been set on the file_name event.
+        self.assertIsNone(fn_event.source.get("time_delta"))
+
+    @mock.patch("timesketch.lib.analyzers.interface.OpenSearchDataStore", MockDataStore)
+    def test_is_suspicious_timestomped(self):
```
Comment (jkppr) [inline timesketch/lib/analyzers/ntfs_timestomp_test.py:156]: Same here, this is also already covered by the `test_is_suspicious` if I'm not mistaken. (case 2 -  It tests the exact same success path and attribute-setting behavior, just with different numbers.)

File: timesketch/lib/analyzers/ntfs_timestomp_test.py [hunk 4]
```diff
@@ -108,9 +108,85 @@ def test_is_suspicious(self):
...
+            file_names=[(fn_event, large_diff)],
+        )
+        self.assertTrue(analyzer.is_suspicious(file_info))
+        # The time_delta attribute must be set on every suspicious file_name.
+        self.assertEqual(fn_event.source.get("time_delta"), large_diff)
+        # The accumulated time_deltas must be stored on the STD_INFO event.
+        self.assertEqual(si_event.source.get("time_deltas"), [large_diff])
+
+    @mock.patch("timesketch.lib.analyzers.interface.OpenSearchDataStore", MockDataStore)
+    def test_is_suspicious_mixed_file_names(self):
```
Comment (jkppr) [inline timesketch/lib/analyzers/ntfs_timestomp_test.py:176]: Same here, this seems also to be already covered by `test_is_suspicious`. (case 3 & 4 - This is the exact same "mixed list with early exit" scenario that the new test is trying to cover.)

File: timesketch/lib/analyzers/domain_test.py [hunk 5]
```diff
@@ -19,3 +32,86 @@ def test_domain_analyzer_class(self):
...
+            "_index": "test",
+            "_source": {
+                "__ts_timeline_id": 1,
+                "timestamp": 1410593222543942,
+                "domain": "rare.example.net",
+            },
+        }
+
+        result = analyzer.run()
+        self.assertIn("2 domains discovered", result)
```
Comment (jkppr) [inline timesketch/lib/analyzers/domain_test.py:96]: Great job setting up `MockDataStore` and running `analyzer.run()`.

Currently, the assertions only check the high-level summary string returned by run() (e.g., "2 domains discovered"). To ensure that the analyzer is actually modifying the events in the datastore as expected (and to prevent future code changes from silently breaking this), we should assert the actual event tags and attributes too.

File: timesketch/lib/datafinder.py [hunk 6]
```diff
@@ -14,13 +14,30 @@
...
 import logging
+from datetime import datetime
 
 from timesketch.lib.analyzers import utils
 from timesketch.lib.datastores.opensearch import OpenSearchDataStore
 
 logger = logging.getLogger("timesketch.data_finder")
 
 
+def _is_valid_iso_date(date_string):
```
Comment (jkppr) [inline timesketch/lib/datafinder.py:25]: That should work well. Would you mind adding a `timesketch/lib/datafinder_test.py` file with some unit tests that verify `set_start_date()` and `set_and_date()` accept valid ISO dates and correctly log warnings for invalid dates?

**Review by jkppr — CHANGES_REQUESTED** [review]
Thanks for the contribution. I have some comments below.
